guice-jpa/src/main/java/com/maxifier/guice/jpa/DBEntityManagerProvider.java

```diff
@@ -0,0 +1,103 @@
+package com.maxifier.guice.jpa;
+
+import com.google.common.reflect.Reflection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+
+/**
+ * Provides @DB context-sensitive {@code EntityManager} and handles it's methods calls.
+ * <p>Create {@code EntityManager} proxy which delegates calls to actual EntityManages from current {@link UnitOfWork}.</p>
+ * <p>Use {@code UnitOfWork} or {@code @DB} to initialize database context.</p>
+ *
+ * @author Konstantin Lyamshin (2015-11-15 21:12)
+ */
+@Singleton
+public class DBEntityManagerProvider implements Provider<EntityManager>, InvocationHandler {
+    private static final Method CLOSE = locateMethodIfExists("close");
+    private static final Method IS_OPEN = locateMethodIfExists("isOpen");
+    private static final Method GET_ENTITY_MANAGER_FACTORY = locateMethodIfExists("getEntityManagerFactory");
+    private static final Method GET_METAMODEL = locateMethodIfExists("getMetamodel");
+    private static final Logger logger = LoggerFactory.getLogger(DBEntityManagerProvider.class);
+    private final EntityManagerFactory entityManagerFactory;
+    private final EntityManager proxy;
+
+    @Inject
+    public DBEntityManagerProvider(EntityManagerFactory entityManagerFactory) {
+        this.entityManagerFactory = entityManagerFactory;
+        this.proxy = Reflection.newProxy(EntityManager.class, this);
+    }
+
+    @Override
+    public EntityManager get() {
+        return proxy;
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+        if (method.getDeclaringClass() == Object.class) {
+            return method.invoke(this, args); // Don't proxy Object's methods
```

Comment:
toString.getDeclarigClass() == Object. toString don't passed to actual EntityManager.